docs/architecture/concentrated-liquidity-adr.md

```diff
@@ -238,19 +261,292 @@ See the next `"Swap Router Module"` section of this document for more details.
+   * Propagate the execution the appropriate module depending on the pool type.
+2. Handle swap messages
+   * Cover & share multihop logic
+   * Propagate intra-pool swaps to the appropriate module depending on the pool type.
+
+Therefore, we move several existing `gamm` messages and tests to the new `swap-router` module,
+connecting them to the `swaprouter` keeper that propagates execution to the appropriate swap module.
+
+The messages to move from `gamm` to `swaprouter` are:
+- `CreatePoolMsg`
+- `MsgSwapExactAmountIn`
+- `MsgSwapExactAmountOut`
+
+Let's consider pool creation and swaps separately and in more detail.
+
+##### Pool Creation & Id Management
 
-Therefore, we move the existing `gamm` swap messages and tests to the new `swap-router` module, connecting to the `swap-router` keeper that simply propagates swaps to `gamm` or `concentrated-liquidity` modules.
+To make sure that the pool ids are unique across the two modules, we unify pool id management
+in the `swaprouter`.
+
+We migrate the store index `next_pool_id` from `gamm` to `swaprouter`. This index represents
+the next available pool id to assign to a newly created pool.
+
+When `CreatePoolMsg` is received, we get the next pool id, assign it to the new pool and propagate
+the execution to either `gamm` or `concentrated-liquidity` modules.
+
+Note that we define a `CreatePoolMsg` interface:
+
+```go
+type CreatePoolMsg interface {
+	// GetPoolType returns the type of the pool to create.
+	GetPoolType() PoolType
+	// The creator of the pool, who pays the PoolCreationFee, provides initial liquidity,
+	// and gets the initial LP shares.
+	PoolCreator() sdk.AccAddress
+	// A stateful validation function.
+	Validate(ctx sdk.Context) error
+	// Initial Liquidity for the pool that the sender is required to send to the pool account
+	InitialLiquidity() sdk.Coins
+	// CreatePool creates a pool implementing PoolI, using data from the message.
+	CreatePool(ctx sdk.Context, poolID uint64) (gammtypes.TraditionalAmmInterface, error)
+}
+```
+
+For each of `balancer`, `stableswap` and `concentrated-liquidity` pools, we have their own implementations.
+
+Let's begin by considering the execution flow of the pool creation message.
+
+1. `CreatePoolMsg` is received by the `swaprouter` message server.
+
+2. `CreatePool` `swaprouter` keeper method is called.
+
+```go
+// x/swaprouter/creator.go CreatePool(...)
+
+// CreatePool attempts to create a pool returning the newly created pool ID or
+// an error upon failure. The pool creation fee is used to fund the community
+// pool. It will create a dedicated module account for the pool and sends the
+// initial liquidity to the created module account.
+//
+// After the initial liquidity is sent to the pool's account, shares are minted
+// and sent to the pool creator. The shares are created using a denomination in
+// the form of gamm/pool/{poolID}. In addition, the x/bank metadata is updated
```

Comment:
this might not be the case for CL pools right?

Reply:
Right, didn't notice. Will update